Feature/add authorize

monsterddq opened the pull request.

Commit:
Add constructor with two parameter: username, passwordDigest for User, Config JWT

Commit:
Update JWT connect to database

Commit:
UPdate

Commit:
Merge branch 'develop' of https://github.com/halo93/e-learning-backend into develop

Commit:
modified:   pom.xml
	modified:   src/main/resources/liquibase/db-changelog.xml

Commit:
Customize AuthenticationException and complete login function

Commit:
Config JWT for authentication + add method convert filed null + add base controller to get CurrentUser

Commit:
Fix/fix error code, format indent, remove class Utility, remove sout

Commit:
Merge branch 'develop' of https://github.com/halo93/e-learning-backend into develop

Commit:
Fix conflict

Commit:
Modify WebSecurityConfig to add feature authorization, Guide in BaseController file, UserController file is simple example

halo93 commented on src/main/resources/liquibase/db-changelog.xml.sample:
```diff
@@ -20,4 +20,5 @@
             insert into articles(article_id, title, category) values(3, 'Van dddhoc', 'Vanhoddddc');
         </sql>
     </changeSet>
```
Xoá cái thẻ mở changset đi

monsterddq replied:
oke

halo93 commented on src/main/java/com/elearningbackend/controller/UserController.java:
```diff
@@ -36,6 +42,7 @@
     }
```
Ô sao thằng này lại mất cái sortby vs direction thế

monsterddq replied:
sao trên project của e vẫn có nhỉ để commit lại

halo93 replied:
OK. Chỗ này a ko nhìn kĩ vào view

Commit:
Fix error in old pull + Customize Access Denie

ChipZero commented on src/main/java/com/elearningbackend/controller/UserController.java:
```diff
@@ -36,6 +42,7 @@
     }
 
     @GetMapping("/users/{key}")
+    @PreAuthorize("hasAuthority('0')")
```
thằng này dùng làm gì đấy cu

monsterddq replied:
đọc comment mới nhất và từ trên xuống dưới a sẽ thấy trong thằng basecontroller

ChipZero commented on src/main/java/com/elearningbackend/customerrorcode/Errors.java:
```diff
@@ -9,7 +9,10 @@
     CATEGORY_NOT_FOUND("008", "CATEGORY NOT FOUND"),
-    SUBCATEGORY_NOT_FOUND("009", "SUBCATEGORY NOT FOUND");
+    SUBCATEGORY_NOT_FOUND("009", "SUBCATEGORY NOT FOUND"),
+    USERNAME_AND_PASSWORD_IS_NOT_EMPTY("010","USERNAME AND PASSWORD IS NOT EMPTY"),
```
thằng này conflig bên anh rồi em ơi! bên answerbank của anh hiện đang  có errors code này rồi

monsterddq replied:
errors code nào ???

ChipZero commented on src/main/java/com/elearningbackend/security/CustomAuthenticationProvider.java:
```diff
@@ -0,0 +1,43 @@
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        String name = authentication.getName().trim();
+        String password = authentication.getCredentials().toString().trim();
+        User user = userRepository.findOne(name);
```
thằng này mình dùng trực tiếp entity ak? có dùng được với dto với mapper không em

monsterddq replied:
dùng được a.
Nhưng với UserDTO là password nhưng đoạn này cần so sánh passwordDigest nền lấy User

ChipZero commented on src/main/java/com/elearningbackend/security/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,35 @@
+                    .setAuthentication(authentication);
+            filterChain.doFilter(request,response);
+        }catch (ElearningException e){
+            ObjectMapper mapper = new ObjectMapper();
```
anh nhớ là mình có thể gọi mapper ra mà không cần phải khai báo một objectMapper mà nhỉ

monsterddq replied:
E không thấy, 
Nếu a định nói trong service a gọi mapper mà k cần khai báo thì a nhìn lại class AbstractService nhé.

halo93 replied:
Ko đc nhé. Sở dĩ a gọi đc trong service là vì em đã khởi tạo ở abstract service rồi

monsterddq replied:
Thì đó, trong server gọi được mapper không cần khai báo do, abstractservice đã khởi tạo rồi, còn class này k kế thừa nên phải khởi tạo riêng. nếu k phải có singleton như kiểu static để dùng 1 đối tượng cho toàn bộ ứng dụng

halo93 replied:
Ok.

ChipZero commented on src/main/java/com/elearningbackend/security/TokenAuthenticationService.java:
```diff
@@ -0,0 +1,76 @@
+            userDto.setAddress(claims.get("address").toString());
+            userDto.setPhone(claims.get("phone").toString());
+            authorities.add(new SimpleGrantedAuthority(userDto.getRole()));
+        }catch (Exception e){
```
nên tạo riêng catch ra vì với token sẽ có các excetion của nó và excetion của mình ! mình tách ra thì mình sẽ thấy rõ error là do đâu

monsterddq replied:
1. Khi trả về client là trả về class result , front-end sẽ chẳng biết exception như thế nào chỉ quan tâm mã lỗi
2. Nên tận dụng những gì đã có để code ngắn gọn hơn.

Commit:
Fix error code ACCESS_DENIED

Commit:
Merge branch 'develop' of https://github.com/halo93/e-learning-backend into develop

Commit:
Fix Conflict

Commit:
modified:   src/main/java/com/elearningbackend/customerrorcode/Errors.java

Commit:
modified:   src/main/java/com/elearningbackend/customerrorcode/Errors.java

Commit:
modified:   src/main/java/com/elearningbackend/customerrorcode/Errors.java
	modified:   src/main/java/com/elearningbackend/security/JWTLoginFilter.java

halo93 commented on README.md:
```diff
@@ -1,8 +0,0 @@
-# e-learning-backend
-Có những rule sau cần ae nhớ:<br>
```
Sao em xóa readme thế?

monsterddq replied:
do e xử lý conflict nên tiện xóa nó để có sự thay đổi , chưa add thêm vào

halo93 replied:
Ok. Em checkout file này ra khỏi pull nhé

Commit:
Merge branch 'develop' of https://github.com/halo93/e-learning-backend into develop

Commit:
Update user activated status + filed activated user entity, user dto + add constants and error code

	new file:   README.md
	modified:   src/main/java/com/elearningbackend/controller/BaseController.java
	modified:   src/main/java/com/elearningbackend/controller/UserController.java
	modified:   src/main/java/com/elearningbackend/customerrorcode/Errors.java
	new file:   src/main/java/com/elearningbackend/dto/CurrentUser.java
	modified:   src/main/java/com/elearningbackend/dto/UserDto.java
	modified:   src/main/java/com/elearningbackend/entity/User.java
	modified:   src/main/java/com/elearningbackend/security/CustomAuthenticationProvider.java
	modified:   src/main/java/com/elearningbackend/security/TokenAuthenticationService.java
	modified:   src/main/java/com/elearningbackend/utility/Constants.java

halo93 commented on src/main/java/com/elearningbackend/controller/BaseController.java:
```diff
@@ -0,0 +1,48 @@
+     */
+    protected CurrentUser getCurrentUser(){
+        authentication = SecurityContextHolder.getContext().getAuthentication();
+        return (CurrentUser) authentication.getPrincipal();
```
Như thế này thì các controller con phải làm cái việc CurrentUser currentUser = getCurrentUser(); 
đúng ko?

monsterddq replied:
đúng thế

halo93 replied:
Để nó toàn cục có đc ko em. Khi cần chỉ cần gọi đến curentUser luôn

monsterddq replied:
toàn cục thì vẫn được, nhưng phải khởi tạo httpcontext thì mới lấy được giá trị, bình thường khi vào strong controller thì context đã được khởi tạo, từ controller trở đi là gọi được. còn trong contructor của controller httpcontext nó vẫn sẽ null.

halo93 replied:
E thử tnay xem đc k
Protected currentuser currentuser = getcurrentuser()

monsterddq replied:
cái này không được đâu, vì khi khởi tạo class sinh ra các filed mà thực hiện luôn hàm này thì bị lỗi null khi complier rồi. Bên C# làm jwt e cũng phải gọi hàm để convert ra user lúc cần lấy.

halo93 replied:
Ok. Thế mình chấp nhận dùng như vậy đi. E check out file readme kia ra nhé. Xong a merge
